### PR TITLE
Add tests to ensure that context is properly reset

### PR DESCRIPTION
**Pull request by jaimergp**
### Description

See https://github.com/conda/conda/pull/13357

### Checklist - did you ...

- [ ] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

**Comment by mbargull** [comment]
I'll also submit a PR to add a test case (using the `add_pip_as_python_dependency: false` as an example we can easily check).

**Comment by jaimergp** [comment]
Awesome! Feel free to push here if you want @mbargull!

**Comment by mbargull** [comment]
Sorry for the commit notification spam :see_no_evil: .

So, the added test does not fail although it should...
I don't actually know why (additional caching someplace else? IDK).

My plan was to add that test with an `xfail` for
`context.solver == "libmamba" and VersionSpec("<FUTURE_FIXED_VERSION").match(conda.__version__)`
in a new PR.
But it's not failing as expected yet :disappointed:.

**Comment by jaimergp** [comment]
https://github.com/conda-forge/stackvana-feedstock/pull/158 does work though :D

**Comment by mbargull** [comment]
> [conda-forge/stackvana-feedstock#158](https://github.com/conda-forge/stackvana-feedstock/pull/158) does work though :D

Oh, yeah, it definitely fixes the issues.

I just want to make sure we have some kind of regression test (esp. since the code around this is pretty finicky).

Now, the issue with the test was that is passed whenever any test has been run before (could have even been the same test for which then only the first invocation failed).
Turns out the code around these `stack_context` functions has little actual functionality and is just made overly complex for what it is -- and, surprise, it's broken.
The `._last_*` members are not properly reset in the `.pop` (& `.replace`) function:
https://github.com/conda/conda/blob/23.10.0/conda/base/context.py#L1861-L1866 .
Hence, `context` is not properly reset.

AFAICT, `conda.plan` is the only consumer of that code.
So, ideally, we can just deprecate and remove that with a short grace period.
Meaning, we don't have to fix it (would actually just be `self._stack[self._stack_idx].apply()` -> `self.apply()`) but can just let it be until removal.

**Comment by jaimergp** [comment]
So we don't need `stack_callback` and can just use `callback` as per https://github.com/conda/conda/pull/13357/commits/c2977d784a71f990c3fa1fd8d9cc31c8f170930c, right? IOW we can close here and mark https://github.com/conda/conda/pull/13357 as ready for review?

**Comment by mbargull** [comment]
> So we don't need `stack_callback` and can just use `callback` as per [conda/conda@c2977d7](https://github.com/conda/conda/commit/c2977d784a71f990c3fa1fd8d9cc31c8f170930c), right?

Yes, much simpler and working :).

> 1. We mark [Use `callback=reset_context` in `conda.plan` conda#13357](https://github.com/conda/conda/pull/13357) as ready for review and wait for it to be merged.

Done.
> 2. We revert the patch in `tests.yml`

Done, and I squashed my debug commit mess and added a `xfail` so the tests will succeed for current and (expected fixed) `conda>23.10.0` future versions.
> 3. We mark this one as ready for review.

Done.

**Comment by jaimergp** [comment]
Hm:

> FAILED tests/test_api_build.py::test_add_pip_as_python_dependency_from_condarc_file[False] -
> Failed: DID NOT RAISE <class 'subprocess.CalledProcessError'>

**Comment by mbargull** [comment]
The canary ones are okay to fail and will pass once the PR for `conda` is merged.
The `22.11.0` one is odd -- I'm taking a look.

**Comment by mbargull** [comment]
Took a look:
For `solver=classic` we of course need the fixed `SubdirData._cache_` mentioned in https://github.com/conda/conda/pull/13357#issuecomment-1824673734 .
I'll add the fix to that PR later today or tomorrow morning.

**Comment by mbargull** [comment]
The tests should fail again now, but succeed whenever https://github.com/conda/conda/pull/13357 is ready and merged.

**Comment by jaimergp** [comment]
Awesome. Thanks @mbargull!

File: tests/test_api_build.py [hunk 1]
```diff
@@ -1899,3 +1902,39 @@ def test_activated_prefixes_in_actual_path(testing_metadata):
...
+    if VersionOrder(CONDA_VERSION) <= VersionOrder("23.10.0"):
+        if not add_pip_as_python_dependency and context.solver == "libmamba":
+            pytest.xfail(
+                "conda.plan.install_actions from conda<=23.10.0 ignores .condarc files."
+            )
+        from conda.base.context import context_stack
+        from conda.core.subdir_data import SubdirData
+
+        # SubdirData's cache doesn't distinguish on add_pip_as_python_dependency.
+        SubdirData.clear_cached_local_channel_data()
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply tests/test_api_build.py:1924]: https://github.com/conda/conda/pull/13357 is in now :D

**Review by jezdez — APPROVED** [review]

**Pull request merged**